src/track/beats.cpp

```diff
@@ -172,14 +214,42 @@ mixxx::BeatsPointer Beats::fromConstTempo(
             lastMarkerPosition.isValid() && lastMarkerBpm.isValid()) {
         return nullptr;
     }
-    return BeatsPointer(new Beats({}, lastMarkerPosition, lastMarkerBpm, sampleRate, subVersion));
+    return BeatsPointer(new Beats({},
+            lastMarkerPosition,
+            lastMarkerBpm,
+            kDefaultBeatsPerBar,
+            sampleRate,
+            subVersion));
+}
+
+bool isDoubleBeat(const QVector<audio::FramePos>& beatPositions, int beatIdx) {
+    if (beatIdx && beatPositions[beatIdx - 1] == beatPositions[beatIdx]) {
+        return true;
+    }
+    return false;
+}
+
+double effectiveHundredthBpm(double beatLength, mixxx::audio::SampleRate sampleRate) {
+    return std::round(60 * sampleRate / beatLength * 100);
 }
 
 // static
 mixxx::BeatsPointer Beats::fromBeatPositions(
         mixxx::audio::SampleRate sampleRate,
         const QVector<audio::FramePos>& beatPositions,
         const QString& subVersion) {
+    // The beat map contains the list of beats contains between the first and
+    // the last beat with custom property (offset, BPM bar length) For example,
+    // you may have the following as input:
+    //
+    // Track: |--------------------------------------------------------|
+    //        |   |   |   |   |   ||  |  |  ||  |  |  ||   |   |   |   ||
```

Comment:
No, this is how the beat marker is translated in the BeatMap: two marker at the exact same positions. By default, the analyser will only issue single beat marker, and so the deserialisation process will infer that the first beat is a beat marker and that the bar length is 4 as default. However, if you set a beat marker (Using the CO, UI available in next PR), it will be serialised using this double beat in the map, the 4-beat bar will be relative to that beat, If you have an other double beat marker withing the acceptable range ([2-32]) it will be used to inferred the bar length, starting from that beat till the end of the track (or the next marker)
I appreciate it may feel a bit complicated, but that was the only way I could reuse the existing proto definition and make that change non-breaking (and I could read it was a point of contention in the past attempt of addressing that feature...)

Reply:
> Will older versions of Mixxx be able to handle the double beats gracefully?

From my testing with 2.4 and 2.5, yes it does. The only issue is a `DEBUG_ASSERT`, but nothing else which seems to impact the correct functionning.

> we might as well add a new "is_marker" field

That would definitely be cleaner. Do we have consnsus on adding a new proto definition? AFAIK, we should be able to add this without issue a new versionning either and only make the most of protobuf field versionning here, so that sounds minimal enought to be acceptable here. Wdyt?

Reply:
> I appreciate it may feel a bit complicated, but that was the only way I could reuse the existing proto definition and make that change non-breaking

I like the idea to have a compatible format. I am just afraid that some sync and quantize code will be confused by the double beat. What are the alternatives? 

```
message BeatMap {
  repeated Beat beat = 1;
}
```

Will an extra array work the bar lines also work? 

```
message BeatMap {
  repeated Beat beat = 1;
  optional repeated Beat bar_line = 2;
}
```

Reply:
I'm wondering if we could just go with this in a first place, and then introduce the new proto definition if it does turns out to break something. After all, that is what alphas are for ;-)